README.md
# ratelimit-middleware [![Build Status](https://travis-ci.org/defunctzombie/ratelimit-middleware.svg?branch=master)](https://travis-ci.org/defunctzombie/ratelimit-middleware)

Rate limit middleware for [expressjs](http://expressjs.com/)

```js
var ratelimit = require('ratelimit-middleware');
var app = express();

app.use(ratelimit({
    burst: 10,  // Max 10 concurrent requests (if tokens)
    rate: 0.5  // Steady state: 1 request / 2 seconds
});

app.get('/throttled/access', function(req, res, next) {
    ...
});
```

## ratelimit(options)

Creates an API rate limiter that can be plugged into the standard
restify request handling pipeline.

This throttle gives you three options on which to throttle:
username, IP address and 'X-Forwarded-For'. IP/XFF is a /32 match,
so keep that in mind if using it.  Username takes the user specified
on req.username (which gets automagically set for supported Authorization
types; otherwise set it yourself with a filter that runs before this).

In both cases, you can set a `burst` and a `rate` (in requests/seconds),
as an integer/float.  Those really translate to the `TokenBucket`
algorithm, so read up on that (or see the comments above...).

In either case, the top level options burst/rate set a blanket throttling
rate, and then you can pass in an `overrides` object with rates for
specific users/IPs.  You should use overrides sparingly, as we make a new
TokenBucket to track each.

### Options

| Name | Default | Type | Description |
| --- | --- | --- | --- |
| rate | - | Number | Steady state number of requests/second to allow |
| burst | - | Number | Amount of requests to burst to |
| ip | true | Boolean | Throttle on /32 (source id) |
| xff | false | Boolean | Throttle on /32 X-Forwarded-For header |
| username | false | Boolean | Throttle on req.username |
| overrides | null | Object | Per "key" overrides |
| tokensTable | - | Object | Storage engine |
| maxKeys | 10000 | Number | Maximum distinct throttling keys to allow at a time |

### Overrides

```js
{
    burst: 10,
    rate: 0.5,
    ip: true,
    overrides: {
        '192.168.1.1': {
            burst: 0,
            rate: 0    // unlimited
        },
        '192.168.1.192/27': {
            burst: 0,
            rate: 0
        }
      }
   }
}
```

## Handle Ratelimit errors

If a request with exceed the rate limit and cannot be processed, the `next` middleware will be invoked with an `Error` argument. The error instance will have a `status` field with code `429` and a `message` indicating the user has exceeded their quota `You have exceeded your request rate of %s r/s.`

You can handle this response by providing error handling middleware in your express app.

```js
var app = express();

app.use(ratelimit({ ... });

app.use(...);
app.get(...);

app.use(function(err, req, res, next) {
   err.status // will be 429 if rate limited

   // example way to respond
   // NOTE, you will likely want to hide `err.message` for 5xx errors in production.
   res.status(err.stauts || 500).send(err.message);
});
```

## Prior art

This module is repackaged code from the [restify](http://restifyjs.com/) library [throttle plugin](https://github.com/mcavage/node-restify/blob/master/lib/plugins/throttle.js) for use with expressjs

## License

MIT

See LICENSE.restify for restify's license.


HISTORY.md
# 1.2.0 (2015-10-05)

  * support multiple ips in x-forwarded-for headers (use first ip)

# 1.1.0 (2015-09-30)

  * add CIDR block parsing to IP based overrides

# 1.0.0 (2015-03-31)

  * initial



tests/xor.js
var assert = require('assert');

var xor = require('../lib/xor');

describe('xor', function() {
    it('should handle T T', function() {
        assert.equal(xor(true, true), false);
    });

    it('should handle F F', function() {
        assert.equal(xor(false, false), false);
    });

    it('should handle T F', function() {
        assert.equal(xor(true, false), true);
    });

    it('should handle F T', function() {
        assert.equal(xor(false, true), true);
    });

    it('should handle T T F', function() {
        assert.equal(xor(true, true, false), false);
    });
});


tests/index.js
var assert = require('assert');
var after = require('after');
var express = require('express');
var supertest = require('supertest');

var ratelimit = require('../');

describe('ratelimit', function() {
    it('should perform basic ratelimiting on ip', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            ip: true
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        app.use(error_handler);

        supertest(app)
        .get('/')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .expect(429)
        .end(function(err, res) {
            assert.equal(res.text, 'You have exceeded your request rate of 1 r/s.');
            done(err);
        });
    });

    it('should perform basic ratelimiting on xff', function(done) {
        done = after(3, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        app.use(error_handler);

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.2')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(429)
        .end(function(err, res) {
            assert.equal(res.text, 'You have exceeded your request rate of 1 r/s.');
            done(err);
        });
    });

    it('should let ip addresses be used as override keys', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true,
            overrides: {
                '1.1.1.1': {
                    burst: 2,
                    rate: 2
                }
            }
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });
    });

    it('should let ip blocks be used as override keys', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true,
            overrides: {
                '1.1.1.192/27': {
                    burst: 2,
                    rate: 2
                }
            }
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.223')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.223')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });
    });

    it('should work when the ip is a comma-delimited list', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true,
            overrides: {
                '1.1.1.192/27': {
                    burst: 2,
                    rate: 2
                }
            }
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.197, 2.2.2.2')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.197, 2.2.2.2')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });
    });

    it('should not apply a block rate limit to an IP that falls outside its range', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true,
            overrides: {
                '1.1.1.192/27': {
                    burst: 2,
                    rate: 2
                }
            }
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        app.use(error_handler);

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', '1.1.1.1')
        .expect(429)
        .end(function(err, res) {
            assert.equal(res.text, 'You have exceeded your request rate of 1 r/s.');
            done(err);
        });
    });

    it('should not fail when the xff is an unexpected value', function(done) {
        done = after(2, done);

        var app = express();
        app.use(ratelimit({
            burst: 1,
            rate: 1,
            xff: true,
            overrides: {
                '1.1.1.192/27': {
                    burst: 2,
                    rate: 2
                }
            }
        }));

        app.get('/', function(req, res, next) {
            res.send('hello');
        });

        app.use(error_handler);

        supertest(app)
        .get('/')
        .set('x-forwarded-for', 'abcxyz')
        .expect(200)
        .end(function(err, res) {
            done(err);
        });

        supertest(app)
        .get('/')
        .set('x-forwarded-for', 'abcxyz')
        .expect(429)
        .end(function(err, res) {
            assert.equal(res.text, 'You have exceeded your request rate of 1 r/s.');
            done(err);
        });
    });
});

function error_handler(err, req, res, next) {
    res.status(err.status || 500);
    res.send(err.message);
};


lib/TokenBucket.js
var assert = require('assert-plus');

/**
 * An implementation of the Token Bucket algorithm.
 *
 * Basically, in network throttling, there are two "mainstream"
 * algorithms for throttling requests, Token Bucket and Leaky Bucket.
 * For restify, I went with Token Bucket.  For a good description of the
 * algorithm, see: http://en.wikipedia.org/wiki/Token_bucket
 *
 * In the options object, you pass in the total tokens and the fill rate.
 * Practically speaking, this means "allow `fill rate` requests/second,
 * with bursts up to `total tokens`".  Note that the bucket is initialized
 * to full.
 *
 * Also, in googling, I came across a concise python implementation, so this
 * is just a port of that. Thanks http://code.activestate.com/recipes/511490 !
 *
 * @param {Object} options contains the parameters:
 *                   - {Number} capacity the maximum burst.
 *                   - {Number} fillRate the rate to refill tokens.
 */
function TokenBucket(options) {
    assert.object(options, 'options');
    assert.number(options.capacity, 'options.capacity');
    assert.number(options.fillRate, 'options.fillRate');

    this.tokens = this.capacity = options.capacity;
    this.fillRate = options.fillRate;
    this.time = Date.now();
}


/**
 * Consume N tokens from the bucket.
 *
 * If there is not capacity, the tokens are not pulled from the bucket.
 *
 * @param {Number} tokens the number of tokens to pull out.
 * @return {Boolean} true if capacity, false otherwise.
 */
TokenBucket.prototype.consume = function consume(tokens) {
    if (tokens <= this._fill()) {
        this.tokens -= tokens;
        return (true);
    }

    return (false);
};


/**
 * Fills the bucket with more tokens.
 *
 * Rather than do some whacky setTimeout() deal, we just approximate refilling
 * the bucket by tracking elapsed time from the last time we touched the bucket.
 *
 * Simply, we set the bucket size to min(totalTokens,
 *                                       current + (fillRate * elapsed time)).
 *
 * @return {Number} the current number of tokens in the bucket.
 */
TokenBucket.prototype._fill = function _fill() {
    var now = Date.now();
    if (now < this.time) // reset account for clock drift (like DST)
        this.time = now - 1000;

    if (this.tokens < this.capacity) {
        var delta = this.fillRate * ((now - this.time) / 1000);
        this.tokens = Math.min(this.capacity, this.tokens + delta);
    }
    this.time = now;

    return (this.tokens);
};

module.exports = TokenBucket;


index.js
var sprintf = require('util').format;
var assert = require('assert-plus');

var xor = require('./lib/xor');
var TokenTable = require('./lib/TokenTable');
var TokenBucket = require('./lib/TokenBucket');
var Netmask = require('netmask').Netmask;

/**
 * Creates an API rate limiter that can be plugged into the standard
 * restify request handling pipeline.
 *
 * This throttle gives you three options on which to throttle:
 * username, IP address and 'X-Forwarded-For'. IP/XFF is a /32 match,
 * so keep that in mind if using it.  Username takes the user specified
 * on req.username (which gets automagically set for supported Authorization
 * types; otherwise set it yourself with a filter that runs before this).
 *
 * In both cases, you can set a `burst` and a `rate` (in requests/seconds),
 * as an integer/float.  Those really translate to the `TokenBucket`
 * algorithm, so read up on that (or see the comments above...).
 *
 * In either case, the top level options burst/rate set a blanket throttling
 * rate, and then you can pass in an `overrides` object with rates for
 * specific users/IPs.  You should use overrides sparingly, as we make a new
 * TokenBucket to track each.
 *
 * On the `options` object ip and username are treated as an XOR.
 *
 * An example options object with overrides:
 *
 *  {
 *    burst: 10,  // Max 10 concurrent requests (if tokens)
 *    rate: 0.5,  // Steady state: 1 request / 2 seconds
 *    ip: true,   // throttle per IP
 *    overrides: {
 *      '192.168.1.1': {
 *        burst: 0,
 *        rate: 0    // unlimited
 *    }
 *  }
 *
 *
 * @param {Object} options required options with:
 *                   - {Number} burst (required).
 *                   - {Number} rate (required).
 *                   - {Boolean} ip (optional).
 *                   - {Boolean} username (optional).
 *                   - {Boolean} xff (optional).
 *                   - {Object} overrides (optional).
 *                   - {Object} tokensTable: a storage engine this plugin will
 *                              use to store throttling keys -> bucket mappings.
 *                              If you don't specify this, the default is to
 *                              use an in-memory O(1) LRU, with 10k distinct
 *                              keys.  Any implementation just needs to support
 *                              put/get.
 *                   - {Number} maxKeys: If using the default implementation,
 *                              you can specify how large you want the table to
 *                              be.  Default is 10000.
 * @return {Function} of type f(req, res, next) to be plugged into a route.
 * @throws {TypeError} on bad input.
 */
function throttle(options) {
    assert.object(options, 'options');
    assert.number(options.burst, 'options.burst');
    assert.number(options.rate, 'options.rate');
    if (!xor(options.ip, options.xff, options.username)) {
        throw new Error('(ip ^ username ^ xff)');
    }

    for (key in options.overrides) {
        var override = options.overrides[key];
        try {
            var block = new Netmask(key);

            // Non-/32 blocks only
            if (block.first !== block.last) {
                override.block = block;
            }
        }
        catch(err) {
            // The key may be a username, which would raise but should
            // be ignored
        }
    }

    var message = options.message || 'You have exceeded your request rate of %s r/s.';

    var table = options.tokensTable || new TokenTable({size: options.maxKeys});

    function rateLimit(req, res, next) {
        var attr;
        var burst = options.burst;
        var rate = options.rate;

        if (options.ip) {
            attr = req.connection.remoteAddress;
        } else if (options.xff) {
            attr = req.headers['x-forwarded-for'];
        } else if (options.username) {
            attr = req.username;
        }

        // Before bothering with overrides, see if this request
        // even matches
        if (!attr) {
            return next(new Error('Invalid throttle configuration'));
        }

        // If the attr is a comma-delimited list of IPs, get the first
        attr = attr.split(',')[0];

        // Check the overrides
        if (options.overrides) {
            var override = options.overrides[attr];

            // If the rate limit attribute matches an override key, apply it
            if (override) {
                if (override.burst !== undefined && override.rate !== undefined) {
                    burst = override.burst;
                    rate = override.rate;
                }
            }

            // Otherwise, see if the rate limit attribute matches any CIDR
            // block overrides
            else {
                for (key in options.overrides) {
                    override = options.overrides[key];
                    var contained = false;

                    try {
                        contained = override.block && override.block.contains(attr);
                    }
                    catch(err) {
                        // attr may be a username, which would raise but should
                        // be ignored
                    }

                    if (contained) {
                        burst = override.burst;
                        rate = override.rate;
                        break;
                    }
                }
            }
        }

        if (!rate || !burst) {
            return next();
        }

        var bucket = table.get(attr);
        if (!bucket) {
            bucket = new TokenBucket({
                capacity: burst,
                fillRate: rate
            });
            table.put(attr, bucket);
        }

        if (!bucket.consume(1)) {
            var msg = sprintf(message, rate);
            var err = new Error(msg);
            err.status = 429; // Too Many Requests
            return next(err);
        }

        return next();
    }

    return rateLimit;
}

module.exports = throttle;
